keras/utils/image_dataset.py

```diff
@@ -261,9 +261,10 @@ def load_image(path, image_size, num_channels, interpolation,
   img = tf.io.read_file(path)
   img = tf.image.decode_image(
       img, channels=num_channels, expand_animations=False)
-  if crop_to_aspect_ratio:
-    img = image_utils.smart_resize(img, image_size, interpolation=interpolation)
-  else:
-    img = tf.image.resize(img, image_size, method=interpolation)
-  img.set_shape((image_size[0], image_size[1], num_channels))
+  if img.shape[0] != image_size[0] or img.shape[1] != image_size[1]:
```

Comment:
Is `img.shape` always statically defined here?

Reply:
`img.shape` gets compared with input image_size before applying  resize method.

Reply:
I mean, are there scenarios where `img.shape[0]` will be None because `tf.image.decode_image()` will return a dynamically sized image? I think this should happen any time the function is called inside a compiled tf.function? Just checking.

Reply:
Each Tensor returned from `tf.image.decode_image()` will have a 3 or 4 dimensional shape, so I don't think it will ever happen that `img.shape[0]` will be `None`. Please take a look at comment [#16227](https://github.com/keras-team/keras/issues/16227#issuecomment-1082934849).

Reply:
Let me explain in more detail then.

A compiled tf.function processes symbolic shapes. When a shape entry can change from sample to sample, it is marked as dynamic, noted as None. Does that make sense?

Try the following code snippet, which is near identical to the code you're editing:

```python
def load_image(path):
  img = tf.io.read_file(path)
  img = tf.image.decode_image(
      img, channels=3, expand_animations=False)
  return img

# Make sure these are real images on disk
image_paths = ['PetImages/Cat/0.jpg', 'PetImages/Cat/1.jpg', 'PetImages/Cat/2.jpg'] 

path_ds = tf.data.Dataset.from_tensor_slices(image_paths)
img_ds = path_ds.map(load_image)

@tf.function
def fn(img_ds):
  for x in img_ds:
    print(x.shape[1])

fn(img_ds)
```

It returns None. So the code you've written would fail in a tf.function.

In general we should never do runtime value checks except in smaller number of special cases, because of the performance overhead of symbolic conditionals.

Reply:
> Perhaps a check for image size should be done in tf.image.resize()

This sounds like the proper solution. Note that it should be runtime check, at the level of the TF op itself, not in Python. It is simply not practical to do such checks in Python.

> I've run your code and it returns me one None, shouldn't it return it 3 times as there are 3 images in dataset?

No, it should return None once because the dataset iterator is symbolic as well. Which is to say that the TF graph does not know in advance how many images are in the dataset, just like it doesn't know what size the images are.

Reply:
As a follow-up please open a feature request in the TensorFlow repo asking for a runtime check in the `tf.image.resize()` op.